app/controllers/users_controller.rb

```diff
@@ -49,9 +50,38 @@ def destroy_session
     head :ok
   end
 
+  def password_recovery
+    @user = User.find_by(username: params[:username])
+    if @user&.email.present?
+      @user.regenerate_password_recovery_token
+      UserMailer.password_recovery(@user).deliver_later
+    end
+    render status: :ok, json: { success: 'The information to reset the password has been sent.' }
+  end
+
+  def password_reset
+    @user = User.find_by(password_recovery_token: params[:password_recovery_token])
+    render 'password_reset_error.html.erb', status: :bad_request if @user.nil?
+  end
+
+  def password_update
+    @user = User.find_by(password_recovery_token: params[:password_recovery_token])
+    if @user&.update(password_params)
+      render 'password_reset_confirmation.html.erb'
+    elsif @user
+      render 'password_reset.html.erb'
+    else
+      render 'password_reset_error.html.erb', status: :bad_request
+    end
+  end
+
   private
 
   def user_params
     params.require(:user).tap { |p| p[:unconfirmed_email] = p[:email] }.permit(:unconfirmed_email, :username, :password)
   end
+
+  def password_params
+    params.require(:user).tap { |p| p[:password_recovery_token] = nil }.permit(:password, :password_confirmation, :password_recovery_token)
```

Comment:
Can this tap be done via the update call instead

i.e. does `@user&.update(password_params, password_recovery_token: nil)` work?